keymanapp/keyman
# fix(web): optimizes, enhances caret placement within text on mobile devices

Fixes #5312.

_Dramatically_ reduces the time taken to place the caret within text for `TouchAliasElement`s, the emulated text inputs KMW produces on mobile devices to facilitate use of the Web-based keyboard.  This is most notable for the first touch into such an area, which had been particularly bad due to some initialization issues for the caret-placement algorithm.

While I was at it, I also fixed up the caret placement logic; it should land where expected far, far more consistently and reliably than before.

# User Testing

This PR adds a new testing page:  `web/testing/issue5312-touch-alias-optimization`.

Said testing page (seen as "Tests performance of touch-alias elements with large amounts of text (#5312)" on the testing page index) features a text area preloaded with 6957 characters.  (Thanks, [Lorem Ipsum Generator](https://loremipsum.io/generator/?n=20&t=p).)  Without the optimizations, my testing tab in Chrome would effectively hang immediately.  (I had to force-quit Chrome a number of times for testing!)  Now... interactions feel nigh-instant!

## Tests

Use remote device emulation on Google Chrome on the testing page noted above.

- GROUP_PHONE:  Run tests on the page while emulating an iPhone SE via Chrome mobile-device emulation.
   - Note that the page will likely auto-scroll on your first touch within the testing text.  (The smaller resolution forces this.)
- GROUP_TABLET:  Run tests on the page while emulating an iPad Air (or similar tablet form-factor) using Chrome mobile-device emulation.

- TEST_INITIAL_TOUCH:  

    - Touch text in the middle of the visible paragraph, preferably toward the center vertically and horizontally.
    - The caret should land very near to the touched location, practically under the touch if possible.
    - Repeat at least 3 times in different locations; ensure it behaves as expected.
    - Finally, repeat once more slightly to the left or right of the previous repetition; ensure it behaves as expected.
    - If _any_ attempt to place a caret via touch for this test takes over 0.25 seconds, **fail** this test.

- TEST_SCROLLED_TOUCH:

    - Scroll the textbox down three paragraphs.
    - Touch text in the middle of the now-top (fourth overall) paragraph, preferably toward the center vertically and horizontally.
    - The caret should land very near to the touched location, practically under the touch if possible.
    - Repeat at least 3 times in different locations; ensure it behaves as expected.
    - Finally, repeat once more slightly to the left or right of the previous repetition; ensure it behaves as expected.
    - If _any_ attempt to place a caret via touch for this test takes over 0.25 seconds, **fail** this test.

As noted, there may be _very slight_ lag on caret-placement attempts, but only very slight.  It shouldn't ever come close to feeling like it's 'locked up' your tab in the browser.

Lone caveat:  unless I'm mistaken, it will not allow you to place the caret at the start of any line but the first.  If it does, it won't allow you to place it at the end of the previous line.  Only one of the two will be possible; that's a `TouchAliasElement` limitation.
